File: backend/app/routers/schema_mapping.py
from typing import Any

from fastapi import APIRouter, HTTPException

from app.schemas.schema_mapping_schema import (
    ApiSchemaCompareRequest,
    ApiSchemaCompareResponse,
    ApiSchemaSummary,
    ApiSchemaTransformPreviewRequest,
    ApiSchemaTransformPreviewResponse,
    InferJsonSchemaRequest,
    InferXmlSchemaRequest,
    MatrixRequest,
    MatrixResponse,
    SchemaCompareRequest,
    SchemaCompareResponse,
    SchemaInferenceResponse,
    SchemaMappingErrorResponse,
    TransformPreviewRequest,
    TransformPreviewResponse,
)
from app.services.schema_mapping.service import (
    SchemaMappingError,
    build_compatibility_matrix,
    compare_schema_pair,
    compare_database_api_schemas,
    infer_json_schema_from_sample,
    infer_xml_schema_from_sample,
    list_database_api_schemas,
    transform_database_mapping_preview,
    transform_preview,
)

router = APIRouter(prefix="/schema-mapping", tags=["schema-mapping"])


def _bad_request(exc: SchemaMappingError) -> HTTPException:
    return HTTPException(status_code=400, detail=str(exc))


@router.post(
    "/compare",
    response_model=SchemaCompareResponse,
    responses={400: {"model": SchemaMappingErrorResponse}},
)
def compare_schemas_endpoint(request: SchemaCompareRequest) -> dict[str, Any]:
    try:
        return compare_schema_pair(
            source_schema=request.source_schema,
            target_schema=request.target_schema,
            source_name=request.source_name,
            target_name=request.target_name,
        )
    except SchemaMappingError as exc:
        raise _bad_request(exc) from exc


@router.get(
    "/api-schemas",
    response_model=list[ApiSchemaSummary],
    responses={400: {"model": SchemaMappingErrorResponse}},
)
def list_api_schemas_endpoint() -> list[dict[str, Any]]:
    try:
        return list_database_api_schemas()
    except SchemaMappingError as exc:
        raise _bad_request(exc) from exc


@router.post(
    "/compare-api-schemas",
    response_model=ApiSchemaCompareResponse,
    responses={400: {"model": SchemaMappingErrorResponse}},
)
def compare_api_schemas_endpoint(request: ApiSchemaCompareRequest) -> dict[str, Any]:
    try:
        return compare_database_api_schemas(
            source_schema_id=request.source_schema_id,
            target_schema_id=request.target_schema_id,
            save_mapping=request.save_mapping,
        )
    except SchemaMappingError as exc:
        raise _bad_request(exc) from exc


@router.post(
    "/transform-api-preview",
    response_model=ApiSchemaTransformPreviewResponse,
    responses={400: {"model": SchemaMappingErrorResponse}},
)
def transform_api_preview_endpoint(request: ApiSchemaTransformPreviewRequest) -> dict[str, Any]:
    try:
        return transform_database_mapping_preview(
            mapping_id=request.mapping_id,
            data=request.data,
            output_format=request.format,
            save_run=request.save_run,
        )
    except SchemaMappingError as exc:
        raise _bad_request(exc) from exc


@router.post(
    "/matrix",
    response_model=MatrixResponse,
    responses={400: {"model": SchemaMappingErrorResponse}},
)
def matrix_endpoint(request: MatrixRequest) -> dict[str, Any]:
    try:
        return build_compatibility_matrix(
            [
                {"name": schema.name, "schema": schema.schema_definition}
                for schema in request.schemas
            ]
        )
    except SchemaMappingError as exc:
        raise _bad_request(exc) from exc


@router.post(
    "/transform-preview",
    response_model=TransformPreviewResponse,
    responses={400: {"model": SchemaMappingErrorResponse}},
)
def transform_preview_endpoint(request: TransformPreviewRequest) -> dict[str, Any]:
    try:
        return transform_preview(
            source_schema=request.source_schema,
            target_schema=request.target_schema,
            data=request.data,
            output_format=request.format,
            source_name=request.source_name,
            target_name=request.target_name,
        )
    except SchemaMappingError as exc:
        raise _bad_request(exc) from exc


@router.post(
    "/infer-json-schema",
    response_model=SchemaInferenceResponse,
    responses={400: {"model": SchemaMappingErrorResponse}},
)
def infer_json_schema_endpoint(request: InferJsonSchemaRequest) -> dict[str, Any]:
    try:
        return {"schema": infer_json_schema_from_sample(request.data)}
    except SchemaMappingError as exc:
        raise _bad_request(exc) from exc


@router.post(
    "/infer-xml-schema",
    response_model=SchemaInferenceResponse,
    responses={400: {"model": SchemaMappingErrorResponse}},
)
def infer_xml_schema_endpoint(request: InferXmlSchemaRequest) -> dict[str, Any]:
    try:
        return {"schema": infer_xml_schema_from_sample(request.xml_content)}
    except SchemaMappingError as exc:
        raise _bad_request(exc) from exc

File: frontend/src/services/schemaMapping.ts
import request from '../utils/request';

export interface SchemaCompareResponse {
  source:             string;
  target:             string;
  compatibility:      string;
  summary:            Record<string, unknown>;
  issues:             Array<Record<string, unknown>>;
  normalized_source:  Record<string, unknown>;
  normalized_target:  Record<string, unknown>;
  [key: string]:      unknown;
}

export interface TransformPreviewResponse {
  result:          unknown;
  format:          'json' | 'xml';
  mapping_status:  string;
  mapping:         Record<string, unknown>;
  comparison:      Record<string, unknown>;
  transform_code:  string;
}

export interface MatrixResponse {
  matrix: Array<Array<Record<string, unknown>>>;
}

export interface SchemaInferenceResponse {
  schema: Record<string, unknown>;
}

export interface ApiSchemaSummary {
  schema_id:          number;
  api_id:             number;
  api_name:           string;
  version_id:         number;
  version_number:     string;
  direction:          string;
  format:             string;
  source_key:         string;
  source_path:        string | null;
  source_method:      string | null;
  media_type:         string | null;
  status_code:        string | null;
  schema_version:     number;
  created_at:         string;
  schema_definition:  Record<string, unknown>;
}

export interface ApiSchemaCompareResponse {
  source_schema:          ApiSchemaSummary;
  target_schema:          ApiSchemaSummary;
  comparison_result_id:   number | null;
  mapping_id:             number | null;
  comparison:             SchemaCompareResponse;
  mapping:                Record<string, unknown>;
}

export interface ApiSchemaTransformPreviewResponse extends TransformPreviewResponse {
  transform_run_id:       number | null;
  comparison_result_id:   number | null;
  mapping_id:             number;
}

/** Schema mapping calls can be slower than CRUD. */
function postMapping<T>(url: string, data: unknown): Promise<T> {
  return request.post(url, data, { timeout: 60000 }) as unknown as Promise<T>;
}

export function compareSchemas(payload: {
  source_schema: Record<string, unknown>;
  target_schema: Record<string, unknown>;
  source_name?:  string;
  target_name?:  string;
}): Promise<SchemaCompareResponse> {
  return postMapping('/api/v1/schema-mapping/compare', payload);
}

export function listApiSchemas(): Promise<ApiSchemaSummary[]> {
  return request.get('/api/v1/schema-mapping/api-schemas') as unknown as Promise<ApiSchemaSummary[]>;
}

export function compareApiSchemas(payload: {
  source_schema_id: number;
  target_schema_id: number;
  save_mapping?:    boolean;
}): Promise<ApiSchemaCompareResponse> {
  return postMapping('/api/v1/schema-mapping/compare-api-schemas', {
    save_mapping: true,
    ...payload,
  });
}

export function transformApiPreview(payload: {
  mapping_id: number;
  data:       Record<string, unknown>;
  format?:    'json' | 'xml';
  save_run?:  boolean;
}): Promise<ApiSchemaTransformPreviewResponse> {
  return postMapping('/api/v1/schema-mapping/transform-api-preview', {
    format: 'json',
    save_run: true,
    ...payload,
  });
}

export function buildMatrix(
  schemas: Array<{ name: string; schema: Record<string, unknown> }>,
): Promise<MatrixResponse> {
  return postMapping('/api/v1/schema-mapping/matrix', { schemas });
}

export function transformPreview(payload: {
  source_schema: Record<string, unknown>;
  target_schema: Record<string, unknown>;
  data:          Record<string, unknown>;
  format?:       'json' | 'xml';
  source_name?:  string;
  target_name?:  string;
}): Promise<TransformPreviewResponse> {
  return postMapping('/api/v1/schema-mapping/transform-preview', {
    format: 'json',
    ...payload,
  });
}

export function inferJsonSchema(data: unknown): Promise<SchemaInferenceResponse> {
  return postMapping('/api/v1/schema-mapping/infer-json-schema', { data });
}

export function inferXmlSchema(xml_content: string): Promise<SchemaInferenceResponse> {
  return postMapping('/api/v1/schema-mapping/infer-xml-schema', { xml_content });
}
